### PR TITLE
[discuss] PATH NEW Dressup TrochoidalProfiling

### PR DESCRIPTION
**Pull request by IchGucksLive**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x ] Branch rebased on latest master `git pull --rebase upstream master`
- [ x] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.

AS SO many Waiting for it to be in the MASTER i got many Requests on YT on this 
 https://www.youtube.com/watch?v=lPrP80Onbog

THE Dressup is Workaroun d till the comunity DECides to get a FULL Trochoidal Operation
on Pocket path edges and faces 
IF this is sceduled for 0.19  this can be used instandly 
HINTS works only on JoinType SQUARE  is internaly Checked and changed on baseOP and recompute
2nd HINT is the Start cand be on a ARC as the Entry move is eFfecting the Angle ERROR Message on ARC start Please USE a straight feed Start point

### CODE REVIEW

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+arccommands = ['G2', 'G3', 'G02', 'G03']
+global currLocation
+currLocation = {}
+
+class ObjectDressup:
+
+    def __init__(self, obj):
+        self.obj = obj
+        obj.addProperty("App::PropertyLink", "Base",  "Path", QtCore.QT_TRANSLATE_NOOP("App::Property", "The base path to modify"))
+        obj.addProperty("App::PropertyDistance", "StepWidth", "Path", QtCore.QT_TRANSLATE_NOOP("App::Property", "Step Alang Path"))
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: I think you mean `Step along path` ?

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+
+    def getHelixDown(self, obj,curCommand):
+        '''returns Helix Down G-code.'''
+        global currLocation
+        results = []
+        op = PathDressup.baseOp(obj.Base)
+        tc = PathDressup.toolController(obj.Base)
+        horizFeed = tc.HorizFeed.Value
+        vertFeed = tc.VertFeed.Value
+        helixradius = op.OffsetExtra.Value - obj.MaterialAlowence.Value
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: obj.Material**Allowance**.Value

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+        return vectorB,vectorA
+
+    def getArcTrochoidal(self,obj,curCommand):
+        '''returns ARC Trochoidal Path G-code.'''
+        global currLocation,test
+        results = []
+        op = PathDressup.baseOp(obj.Base)
+        tc = PathDressup.toolController(obj.Base)
+        horizFeed = tc.HorizFeed.Value
+        helixradius = op.OffsetExtra.Value - obj.MaterialAlowence.Value
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: obj.Material**Allowance**.Value

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+
+    def getTrochoidal(self, obj,curCommand):
+        '''returns Trochoidal Path G-code.'''
+        global currLocation
+        #PathLog.info("HIER Straight {}".format(obj.StepWidth))
+        results = []
+        op = PathDressup.baseOp(obj.Base)
+        tc = PathDressup.toolController(obj.Base)
+        horizFeed = tc.HorizFeed.Value
+        helixradius = op.OffsetExtra.Value - obj.MaterialAlowence.Value
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: obj.Material**Allowance**.Value

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+global currLocation
+currLocation = {}
+
+class ObjectDressup:
+
+    def __init__(self, obj):
+        self.obj = obj
+        obj.addProperty("App::PropertyLink", "Base",  "Path", QtCore.QT_TRANSLATE_NOOP("App::Property", "The base path to modify"))
+        obj.addProperty("App::PropertyDistance", "StepWidth", "Path", QtCore.QT_TRANSLATE_NOOP("App::Property", "Step Alang Path"))
+        obj.addProperty("App::PropertyDistance", "MaterialAlowence", "Path", QtCore.QT_TRANSLATE_NOOP("App::Property", "Keep Material to Finish"))
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: Should be Material**Allowance**
also 
is this correct: `"Keep Material to Finish"` ?

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+
+    def setup(self, obj):
+        
+        tc = PathDressup.toolController(obj.Base)
+        if hasattr(tc, 'Tool'):
+            obj.StepWidth = tc.Tool.Diameter / 10
+        else:
+            PathLog.error(translate("NO TOOL detected \n")) 
+            obj.StepWidth = 0.5           
+        obj.MaterialAlowence = 0.0
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: obj.Material**Allowance**

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+        firstmove = Path.Command("G0", {"X": 0, "Y": 0, "Z": 0})
+        currLocation.update(firstmove.Parameters)
+        newpath = []
+        firstMayBeArc = True
+        for curCommand in obj.Base.Path.Commands:
+            # don't worry about non-move commands, just add to output
+            if curCommand.Name not in movecommands + rapidcommands:
+                newpath.append(curCommand)
+                continue
+            # rapid move straigt no action
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: rapid move **straight** no action

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+            if curCommand.Name in rapidcommands:
+                newpath.append(curCommand)
+                currLocation.update(curCommand.Parameters)
+                continue
+            # FEED Move TAKE action
+            if curCommand.Name in movecommands:
+                if curCommand.z != currLocation["Z"]:#moviung DOWN
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: #**moving** DOWN

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+            extendcommand = Path.Command('G1', {"X": curCommand.x-helixradius, "Y": curCommand.y, "Z": currLocation["Z"], "F": horizFeed})
+            results.append(extendcommand)
+            momentaryZ= currLocation["Z"]
+            while momentaryZ-obj.StepWidth.Value > curCommand.z:
+                momentaryZ = momentaryZ-obj.StepWidth.Value
+                arcmove = Path.Command('G2', {"X": curCommand.x-helixradius, "Y": curCommand.y, "I": helixradius,"Z":momentaryZ, "F": vertFeed})  # add G2/G3 move
+                results.append(arcmove)
+            arcmove = Path.Command('G2', {"X": curCommand.x-helixradius, "Y": curCommand.y, "I": helixradius,"Z":curCommand.z, "F": vertFeed})  # add G2/G3 move
+            results.append(arcmove)
+            #results.append(arcmove) #full turn on bottem
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: #results.append(arcmove) #full turn on **bottom**

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+            extendcommand = Path.Command('G1', {"X": curCommand.x+helixradius, "Y": curCommand.y, "Z": currLocation["Z"], "F": horizFeed})
+            results.append(extendcommand)
+            momentaryZ= currLocation["Z"]
+            while momentaryZ-obj.StepWidth.Value > curCommand.z:
+                momentaryZ = momentaryZ-obj.StepWidth.Value
+                arcmove = Path.Command('G3', {"X": curCommand.x+helixradius, "Y": curCommand.y, "I": helixradius*-1 ,"Z":momentaryZ, "F": vertFeed})  # add G2/G3 move
+                results.append(arcmove)
+            arcmove = Path.Command('G3', {"X": curCommand.x+helixradius, "Y": curCommand.y, "I": helixradius*-1,"Z":curCommand.z, "F": vertFeed})  # add G2/G3 move
+            results.append(arcmove)
+            #results.append(arcmove) #full turn on bottem
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: #results.append(arcmove) #full turn on **bottom**

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+            arcdir= 'CW'
+        p2 = curCommand.Placement.Base
+        edge=PathGeom.edgeForCmd(curCommand,currendPos)
+        arclength = edge.Length # length of the Circle Segment
+        #PathLog.info("arclength {} ".format(arclength))
+        arccenter = FreeCAD.Vector(currendPos.x + curCommand.I, currendPos.y + curCommand.J, currendPos.z)
+        #PathLog.info("arccenter X{} Y{}".format(arccenter.x,arccenter.y))
+        radius = math.hypot(curCommand.I, curCommand.J)
+        #PathLog.info("radius {} ".format(radius))
+        #Data from circle command / Angle includet and direction Start End Angle
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: #Data from circle command / Angle **included** and direction Start End Angle

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+        anglelineCenterEnd = D.angle(curCommand.Placement.Base.sub(arccenter), FreeCAD.Base.Vector(1, 0, 0), FreeCAD.Base.Vector(0, 0, -1))
+        #PathLog.info("anglelineCenterEnd deg {} ".format(math.degrees(anglelineCenterEnd)))
+        anglelineCenterStart = D.angle(currendPos.sub(arccenter), FreeCAD.Base.Vector(1, 0, 0), FreeCAD.Base.Vector(0, 0, -1))
+        arcangle=PathGeom.diffAngle(anglelineCenterEnd,anglelineStartCenter,arcdir)
+        #arcangle = anglelineCenterStart+anglelineCenterEnd
+        #PathLog.info("arcangle deg {} ".format(math.degrees(arcangle)))
+        #PathLog.info("anglelineCenterStart deg {} ".format(math.degrees(anglelineCenterStart)))
+        #neuarcangle= anglelineCenterStart+anglelineCenterEnd
+        #PathLog.info("neuarcangle deg {} ".format(math.degrees(neuarcangle)))
+        stepsAlongCircle = int(abs(arclength/obj.StepWidth.Value))#howmany steps to the circle
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: #**how many** steps to the circle

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+    def getTrochoidal(self, obj,curCommand):
+        '''returns Trochoidal Path G-code.'''
+        global currLocation
+        #PathLog.info("HIER Straight {}".format(obj.StepWidth))
+        results = []
+        op = PathDressup.baseOp(obj.Base)
+        tc = PathDressup.toolController(obj.Base)
+        horizFeed = tc.HorizFeed.Value
+        helixradius = op.OffsetExtra.Value - obj.MaterialAlowence.Value
+        #calulate points
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: #**calculate** points

File: src/Mod/Path/PathScripts/PathDressupTrochoidal.py [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+import PathScripts.PathLog as PathLog
+import PathScripts.PathUtils as PathUtils
+import math
+import DraftVecUtils as D
+from PySide import QtCore
+from PathScripts.PathGeom import PathGeom
+
+"""Trochoidal Milling Path Dressup to profile"""
+
+# Qt tanslation handling
```
Comment (luzpaz) [inline src/Mod/Path/PathScripts/PathDressupTrochoidal.py]: Qt **translation** handling

**Comment by IchGucksLive** [comment]
@luzpaz  correct it is a start and as we are on this since weeks to 0.17 release i am also with  sliptronics discussion to get this as separated function 
BUT as we all are non pro coders and for me i am not native English as you noticed 
AS well as i am not familiar with the system GIT Do i need to delete the PR to change your guess 

I AM a very Old Man and my thinking is quite poor now i relay got sometimes in trouble with all this calculations as well as the FC system module BackandForth operation 

Please See it as a Heavy development Forward operation  Need Further Delete

**Pull request closed without merging**